### PR TITLE
Prometheus: Fill missing steps with null values

### LINKED ISSUE #42993: Null values always connected in time series panel in 8.3.1
**What happened**:
When using a time series panel to visualise Prometheus time series data using a line style graph, when there are null values (i.e. no data retrieved from Prometheus for a period), the line connects these values. This behaviour has only occured since upgrading from Grafana 8.2.7 to 8.3.1.
Changing the selected option Graph styles/connect null values from Never, to Always, to Threshold has no impact.

The panel in Grafana 8.2.7:
Screenshot 2021-12-10 at 16 56 17

The panel in Grafana 8.3.1:
Screenshot 2021-12-10 at 18 12 05

**What you expected to happen**:
I expected that the line would stop and not be shown when there is no data.

**How to reproduce it (as minimally and precisely as possible)**:
In Grafana 8.3.1, create a time series panel with a line graph style using a query containing null values. Try changing the 'Connect null values' option in graph styles and see that it never makes a difference. 

**Anything else we need to know?**:
[This](https://github.com/grafana/grafana/issues/40698) is a related issue about the same problem in the old Graph panel, which was solved by this [PR](https://github.com/grafana/grafana/pull/40757). However, I've tried changing the Time Series panel to the old Graph panel, and changing the null value options to null or null as zero doesn't work, the points are always connected.

**Environment**:
- Grafana version: 8.3.1
- Data source type & version: Prometheus 
- OS Grafana is installed on: 
- User OS & Browser: macOS Catalina/Firefox
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:

Previously (before backend migration) on [frontend we filled out missing steps with null values](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/result_transformer.ts#L383-L389). This PR implements this missing functionality. I've decided to implement this on backend so the results of prometheus queries are same everywhere (Dashboard, Explore, Alerting) and we don't have to rely o frontend implementation. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/42993

**How to test:**
1. Run make `devenv sources=prometheus`
2. Create dashboard with query `counters_logins{server="backend-01"}` and time range `Last 5 min` and step `10s`
3. Stop `devenv_fake-prometheus-data_1` in docker and wait ~1 min. Then start it again
4. Where data points are missing, you should see empty/non-connected space:
image

On current main, the lines are connected (incorrect behaviour/bug):
image

### CODE REVIEW

File: pkg/tsdb/prometheus/time_series_query.go [hunk 1]
```diff
@@ -73,8 +73,8 @@ func (s *Service) executeTimeSeriesQuery(ctx context.Context, req *backend.Query
 		timeRange := apiv1.Range{
 			Step: query.Step,
 			// Align query range to step. It rounds start and end down to a multiple of step.
-			Start: time.Unix(int64(math.Floor((float64(query.Start.Unix()+query.UtcOffsetSec)/query.Step.Seconds()))*query.Step.Seconds()-float64(query.UtcOffsetSec)), 0),
-			End:   time.Unix(int64(math.Floor((float64(query.End.Unix()+query.UtcOffsetSec)/query.Step.Seconds()))*query.Step.Seconds()-float64(query.UtcOffsetSec)), 0),
+			Start: alignTimeRange(query.Start, query.Step, query.UtcOffsetSec),
```
Comment (ivanahuckova) [inline pkg/tsdb/prometheus/time_series_query.go:76]: Just a refactoring. As we want to use aligning at multiple places, we created function for aligning.

File: pkg/tsdb/prometheus/time_series_query.go [hunk 2]
```diff
@@ -299,15 +299,28 @@ func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data
 			tags[string(k)] = string(v)
 		}
 
-		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, len(v.Values))
-		valueField := data.NewFieldFromFieldType(data.FieldTypeNullableFloat64, len(v.Values))
+		baseTimestamp := alignTimeRange(query.Start, query.Step, query.UtcOffsetSec).UnixMilli()
+		datapointsCount := int((query.End.UnixMilli()-query.Start.UnixMilli())/query.Step.Milliseconds()) + 1
```
Comment (ivanahuckova) [inline pkg/tsdb/prometheus/time_series_query.go:303]: Number of data points that we use in`NewFieldFromFieldType` as we no longer can use `len(v.Values)`.

File: pkg/tsdb/prometheus/time_series_query.go [hunk 3]
```diff
@@ -299,15 +299,28 @@ func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data
...
 
-		for i, k := range v.Values {
-			timeField.Set(i, time.Unix(k.Timestamp.Unix(), 0).UTC())
-			value := float64(k.Value)
+		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, datapointsCount)
+		valueField := data.NewFieldFromFieldType(data.FieldTypeNullableFloat64, datapointsCount)
+		idx := 0
```
Comment (ivanahuckova) [inline pkg/tsdb/prometheus/time_series_query.go:307]: Also here, we can't rely on index from `for _, pair := range v.Values {` and we have to keep the track ourself.

File: pkg/tsdb/prometheus/time_series_query.go [hunk 4]
```diff
@@ -299,15 +299,28 @@ func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data
...
+
+		for _, pair := range v.Values {
+			timestamp := int64(pair.Timestamp)
+			value := float64(pair.Value)
+
+			for t := baseTimestamp; t < timestamp; t += query.Step.Milliseconds() {
+				timeField.Set(idx, time.Unix(0, t*1000000).UTC())
```
Comment (ivanahuckova) [inline pkg/tsdb/prometheus/time_series_query.go:314]: If we set no value in valueField for that index, it default to nil (which is what we want).

File: pkg/tsdb/prometheus/time_series_query.go (hunk 1 above)
Comment (MasslessParticle) [reply pkg/tsdb/prometheus/time_series_query.go:76]: This is an awesome improvement.

File: pkg/tsdb/prometheus/time_series_query.go (hunk 3 above)
Comment (MasslessParticle) [reply pkg/tsdb/prometheus/time_series_query.go:309]: This all looks good. 

Nit: Now that there's more going on, pulling this out to a function that returns `timeField, valueField := ...` might read a little easier.

File: pkg/tsdb/prometheus/time_series_query_test.go [hunk 5]
```diff
@@ -586,6 +590,38 @@ func TestPrometheus_parseTimeSeriesResponse(t *testing.T) {
...
+		res, err := parseTimeSeriesResponse(value, query)
+		var nilPointer *float64
+
+		require.NoError(t, err)
+		require.Len(t, res, 1)
+		require.Equal(t, res[0].Fields[0].Len(), 4)
+		require.Equal(t, res[0].Fields[0].At(1), time.Unix(2, 0).UTC())
+		require.Equal(t, res[0].Fields[0].At(2), time.Unix(3, 0).UTC())
+		require.Equal(t, res[0].Fields[1].Len(), 4)
+		require.Equal(t, res[0].Fields[1].At(1), nilPointer)
```
Comment (MasslessParticle) [inline pkg/tsdb/prometheus/time_series_query_test.go:621]: Nit: These can just be `require.Nil(t, ...)`

**Review by MasslessParticle — APPROVED** [review]
Looks good. I think we can improve the readability just a bit.

File: pkg/tsdb/prometheus/time_series_query.go [hunk 6]
```diff
@@ -299,15 +299,30 @@ func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data
...
+				idx++
+			}
+
+			timeField.Set(idx, time.Unix(pair.Timestamp.Unix(), 0).UTC())
```
Comment (toddtreece) [inline pkg/tsdb/prometheus/time_series_query.go:320]: i'm not sure of the best way to phrase this question, so let me know if this doesn't make sense: is it safe to assume that `pair.Timestamp.Unix()` will match the step spacing of the loop above?

**Review by toddtreece — APPROVED** [review]

File: pkg/tsdb/prometheus/time_series_query.go (hunk 6 above)
Comment (ivanahuckova) [reply pkg/tsdb/prometheus/time_series_query.go:320]: Yeah, good question! I would say it is safe as we were [doing the same thing on frontend](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/result_transformer.ts#L383-L389). Moreover, that's how Prometheus returns data - for each step 1 data point and they are exactly 1 step spaced.

**Comment by PayBas** [comment]
@ivanahuckova is this the new intended behavior with stacking?

image

image

image

**Comment by ivanahuckova** [comment]
@PayBas yes this is a correct behaviour. Before the 8.3 there were gaps but in 8.3. we have introduced bug that removed gaps. So 8.3.4 brings it back. 🙂

**Comment by PayBas** [comment]
@ivanahuckova thank you for your reply. I am a little confused however. The bottom 2 images (which are the same graph/panel) shows that "Connect null values: Always" will correctly show uninterrupted lines, whereas turning "Stack series" to either "Normal" or "100%" will introduce gaps.

I'm concerned this PR has had an unexpected side-effect on the stacking rendering. Shouldn't stacked graphs render uninterrupted when using "Connect null values: Always".

What I'm saying is: my panel settings were showing correctly all the way from `7.0.0` right until `8.3.4`.

**Comment by ivanahuckova** [comment]
@PayBas yes, might be a regression. Could you please create separate issue for this. Thank you!

**Pull request merged**